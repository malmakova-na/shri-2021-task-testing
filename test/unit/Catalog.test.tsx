import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import events from "@testing-library/user-event";
import { Router } from "react-router";
import { Provider } from "react-redux";
import { createMemoryHistory } from 'history';
import renderer from 'react-test-renderer';

import { Application } from "../../src/client/Application";
import { CartApi } from "../../src/client/api";
import { initStore, productsLoad, productsLoaded, productDetailsLoad, productDetailsLoaded } from "../../src/client/store";
import { TestExampleApi } from "./TestApi";
import { PRODUCTS, getProductById, CartData, CartData_Test_1, CartData_Test_2, CartData_Test_3 } from './TestData';

describe("Catalog: ", function (){
    //console.log("store ",TestExampleStore)
    const basename = "/hw/store";
    const history = createMemoryHistory({
        initialEntries: ["/catalog"],
        initialIndex: 0
    });
    const api = new TestExampleApi(basename);
    const cart = new CartApi();
    //cart.setState(CartData);
    const TestExampleStore = initStore(api, cart);
    const application = (
        <Router history={history}>
            <Provider store={TestExampleStore}>
            <Application />
            </Provider>
        </Router>
    );

    it("тест на отображение товаров в каталоге: ", () => {
    
        const { container} = render(application);
        
        TestExampleStore.dispatch(productsLoad());
        expect(screen.getByText(/loading/i).innerHTML).toBe('LOADING');
        TestExampleStore.dispatch(productsLoaded(PRODUCTS));
        //screen.logTestingPlaygroundURL();

        
        const imgs = container.querySelectorAll('img.Image');
        expect(imgs.length).toBe(PRODUCTS.length);
        //screen.logTestingPlaygroundURL()


    
        const products = container.querySelectorAll(".card-body");
        //console.log("cart state", TestExampleStore.getState())
        expect(products.length).toEqual(PRODUCTS.length);//правильное количество товаров
        products.forEach((product, id) => {
            const testExample = getProductById(id);
            //правильное название товара
            const name = product.querySelector('h5.ProductItem-Name.card-title').innerHTML;
            expect(name).toBe(testExample.name);
            //правильное цена
            const price = product.querySelector('p.ProductItem-Price.card-text').innerHTML;
            expect(price).toBe('$' + testExample.price);
            //правильная ссылка
            const link = product.querySelector('a.ProductItem-DetailsLink.card-link').getAttribute('href');
            expect(link).toBe(`/catalog/${id}`);
        });
        //const productStatus = container.querySelector('span.CartBadge');
        //expect(productStatus).toBeNull();
        expect(screen.getByRole('link', {
            name: /cart/i
        })).toBeInTheDocument();
        //screen.logTestingPlaygroundURL();
       
        
    });

    it("тест на отображение деталей о товаре ", () => {
        const history = createMemoryHistory({
            initialEntries: ["/catalog"],
            initialIndex: 0
        });
        const api = new TestExampleApi(basename);
        const cart = new CartApi();
        const TestExampleStore = initStore(api, cart);
        const application = (
            <Router history={history}>
                <Provider store={TestExampleStore}>
                <Application />
                </Provider>
            </Router>
        );
        const { container, getByRole} = render(application);

        const testId = 0;
        const product = getProductById(testId);

        TestExampleStore.dispatch(productsLoad());
        TestExampleStore.dispatch(productsLoaded(PRODUCTS));

        const detailsLink = container.querySelector('.ProductItem-DetailsLink');
        expect(detailsLink).toBeInTheDocument();
        events.click(detailsLink); 
        
        TestExampleStore.dispatch(productDetailsLoad(product.id));
        TestExampleStore.dispatch(productDetailsLoaded(product));

        const name = container.querySelector('h1.ProductDetails-Name').innerHTML;
        expect(name).toBe(product.name);

        const description = container.querySelector('p.ProductDetails-Description').innerHTML;
        expect(description).toBe(product.description);

        const price = container.querySelector('p.ProductDetails-Price').innerHTML;
        expect(price).toBe('$' + product.price);

        const color = container.querySelector('dd.ProductDetails-Color').innerHTML;
        expect(color).toBe(product.color);

        const material = container.querySelector('dd.ProductDetails-Material').innerHTML;
        expect(material).toBe(product.material);

        const img = container.querySelector('.Image');
        expect(img).toBeInTheDocument();
        //screen.logTestingPlaygroundURL()

        expect(
            screen.getByRole('button', {
                name: /add to cart/i
            })
        ).toBeInTheDocument();

        const productSnapshot = renderer.create(application).toJSON();
      
        expect(productSnapshot).toMatchSnapshot();
        
        //screen.logTestingPlaygroundURL()
    });
    
    it("тест на добавление одного товара в корзину ", () => {
        const basename = "/hw/store";
        const history = createMemoryHistory({
            initialEntries: ["/catalog"],
            initialIndex: 0
        });
        const api = new TestExampleApi(basename);
        const cart = new CartApi();
        cart.setState(CartData_Test_1);
        const TestExampleStore = initStore(api, cart);
        const application = (
            <Router history={history}>
                <Provider store={TestExampleStore}>
                <Application />
                </Provider>
            </Router>
        );
        const { container, getByRole} = render(application);
        TestExampleStore.dispatch(productsLoaded(PRODUCTS));
        //console.log(TestExampleStore.getState())
        expect(screen.getByText(/item in cart/i)).toBeInTheDocument();
        expect(screen.getByRole('link', {
            name: /cart \(1\)/i
        })).toBeInTheDocument();

        //screen.logTestingPlaygroundURL()
        //expect(screen.getByText(/item in cart/i)).toBeInTheDocument();        
        
    });
    it("тест на пустую корзину ", () => {
        const { container, getByRole} = render(application);
        TestExampleStore.dispatch(productsLoaded(PRODUCTS));
        expect(screen.getByRole('link', {
            name: /cart/i
        })).toBeInTheDocument();  
        const cardBarges = container.querySelectorAll("span.CartBadge");
        expect(cardBarges.length).toBe(0);
        const cartBtn = screen.getByRole('link', {
            name: /cart/i
        });
        events.click(cartBtn);
        expect(screen.getByText(/cart is empty\. please select products in the \./i)).toBeInTheDocument();
        const emptyCartSnapshot = renderer.create(application).toJSON();
      
        expect(emptyCartSnapshot).toMatchSnapshot();
       // screen.logTestingPlaygroundURL()
        //screen.logTestingPlaygroundURL()     
    });

    it("тест на добавление одинакавых товаров  в корзину ", () => {
        const basename = "/hw/store";
        const history = createMemoryHistory({
            initialEntries: ["/catalog"],
            initialIndex: 0
        });
        const api = new TestExampleApi(basename);
        const cart = new CartApi();
        cart.setState(CartData_Test_2);
        const TestExampleStore = initStore(api, cart);
        const application = (
            <Router history={history}>
                <Provider store={TestExampleStore}>
                <Application />
                </Provider>
            </Router>
        );
        const { container, getByRole} = render(application);
        TestExampleStore.dispatch(productsLoaded(PRODUCTS));
        expect(screen.getByRole('link', {
            name: /cart \(1\)/i
        })).toBeInTheDocument();
        expect(screen.getByText(/item in cart/i)).toBeInTheDocument();
        expect(container.textContent).toContain("Item in cart");
        const cardBarges = container.querySelectorAll("span.CartBadge");
        expect(cardBarges.length).toBe(1);
        //screen.logTestingPlaygroundURL()
        
    });

    it("тест на добавление разных товаров  в корзину ", () => {
        const basename = "/hw/store";
        const history = createMemoryHistory({
            initialEntries: ["/catalog"],
            initialIndex: 0
        });
        const api = new TestExampleApi(basename);
        const cart = new CartApi();
        cart.setState(CartData_Test_3);
        const TestExampleStore = initStore(api, cart);
        const application = (
            <Router history={history}>
                <Provider store={TestExampleStore}>
                <Application />
                </Provider>
            </Router>
        );
        const { container, getByRole} = render(application);
        TestExampleStore.dispatch(productsLoaded(PRODUCTS));
        expect(screen.getByRole('link', {
            name: /cart \(2\)/i
        })).toBeInTheDocument();
        const cardBarges = container.querySelectorAll("span.CartBadge");
        expect(cardBarges.length).toBe(2);
        //expect(container.textContent).toContain("Item in cart");
        //screen.logTestingPlaygroundURL()
        
    });

    it("тест на  отображение корзины ", () => {
        const basename = "/hw/store";
        const history = createMemoryHistory({
            initialEntries: ["/cart"],
            initialIndex: 0
        });
        const api = new TestExampleApi(basename);
        const cart = new CartApi();
        cart.setState(CartData_Test_1);
        const TestExampleStore = initStore(api, cart);
        const application = (
            <Router history={history}>
                <Provider store={TestExampleStore}>
                <Application />
                </Provider>
            </Router>
        );
        const { container, getByRole} = render(application);
        TestExampleStore.dispatch(productsLoaded(PRODUCTS));
        //console.log(TestExampleStore.getState())
        expect(screen.getByRole('table')).toBeInTheDocument();
        const cartSnapshot = renderer.create(application).toJSON();
      
        expect(cartSnapshot).toMatchSnapshot();
        //screen.logTestingPlaygroundURL()
        //expect(screen.getByText(/item in cart/i)).toBeInTheDocument();        
        
    });
});