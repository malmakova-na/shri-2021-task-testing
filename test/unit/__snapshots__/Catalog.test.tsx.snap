// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalog:  тест на  отображение корзины  1`] = `
<div
  className="Application"
>
  <nav
    className="navbar navbar-expand-sm navbar-light bg-light"
  >
    <div
      className="container"
    >
      <a
        className="Application-Brand navbar-brand"
        href="/"
        onClick={[Function]}
      >
        Example store
      </a>
      <button
        aria-label="Toggle navigation"
        className="Application-Toggler navbar-toggler"
        onClick={[Function]}
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <div
        className="Application-Menu collapse navbar-collapse"
      >
        <div
          className="navbar-nav"
        >
          <a
            aria-current={null}
            className="nav-link"
            href="/catalog"
            onClick={[Function]}
          >
            Catalog
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/delivery"
            onClick={[Function]}
          >
            Delivery
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/contacts"
            onClick={[Function]}
          >
            Contacts
          </a>
          <a
            aria-current="page"
            className="nav-link active"
            href="/cart"
            onClick={[Function]}
            style={Object {}}
          >
            Cart (1)
          </a>
        </div>
      </div>
    </div>
  </nav>
  <div
    className="container pt-4"
  >
    <div
      className="Cart"
    >
      <div
        className="row mb-4"
      >
        <div
          className="col"
        >
          <h1>
            Shopping cart
          </h1>
          <table
            className="Cart-Table table"
          >
            <thead>
              <tr>
                <th
                  scope="col"
                >
                  #
                </th>
                <th
                  scope="col"
                >
                  Product
                </th>
                <th
                  scope="col"
                >
                  Price
                </th>
                <th
                  scope="col"
                >
                  Count
                </th>
                <th
                  scope="col"
                >
                  Total
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                data-testid="0"
              >
                <th
                  className="Cart-Index"
                  scope="row"
                >
                  1
                </th>
                <td
                  className="Cart-Name"
                >
                  test-product-name
                </td>
                <td
                  className="Cart-Price"
                >
                  $
                  1
                </td>
                <td
                  className="Cart-Count"
                >
                  1
                </td>
                <td
                  className="Cart-Total"
                >
                  $
                  1
                </td>
              </tr>
            </tbody>
            <tfoot>
              <tr>
                <td
                  colSpan={4}
                >
                  Order price:
                </td>
                <td
                  className="Cart-OrderPrice"
                >
                  $
                  1
                </td>
              </tr>
            </tfoot>
          </table>
        </div>
      </div>
      <div
        className="row mb-4"
      >
        <div
          className="col-6"
        >
          <button
            className="Cart-Clear btn btn-outline-secondary"
            onClick={[Function]}
          >
            Clear shopping cart
          </button>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-12 col-sm-8 col-md-6 col-lg-4"
        >
          <h2>
            Сheckout
          </h2>
          <div
            className="Form"
          >
            <div
              className="mb-3"
            >
              <label
                className="form-label"
                htmlFor="f-name"
              >
                Name
              </label>
              <input
                autoComplete="off"
                className="Form-Field Form-Field_type_name form-control"
                disabled={false}
                id="f-name"
                onChange={[Function]}
                type="text"
              />
              <div
                className="invalid-feedback"
              >
                Please provide your name
              </div>
            </div>
            <div
              className="mb-3"
            >
              <label
                className="form-label"
                htmlFor="f-phone"
              >
                Phone
              </label>
              <input
                className="Form-Field Form-Field_type_phone form-control"
                disabled={false}
                id="f-phone"
                onChange={[Function]}
                type="text"
              />
              <div
                className="invalid-feedback"
              >
                Please provide a valid phone
              </div>
            </div>
            <div
              className="mb-3"
            >
              <label
                className="form-label"
                htmlFor="f-address"
              >
                Address
              </label>
              <textarea
                className="Form-Field Form-Field_type_address form-control"
                disabled={false}
                id="f-address"
                onChange={[Function]}
                rows={3}
              />
              <div
                className="invalid-feedback"
              >
                Please provide a valid address
              </div>
            </div>
            <button
              className="Form-Submit btn btn-primary"
              disabled={false}
              onClick={[Function]}
            >
              Checkout
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Catalog:  тест на отображение деталей о товаре  1`] = `
<div
  className="Application"
>
  <nav
    className="navbar navbar-expand-sm navbar-light bg-light"
  >
    <div
      className="container"
    >
      <a
        className="Application-Brand navbar-brand"
        href="/"
        onClick={[Function]}
      >
        Example store
      </a>
      <button
        aria-label="Toggle navigation"
        className="Application-Toggler navbar-toggler"
        onClick={[Function]}
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <div
        className="Application-Menu collapse navbar-collapse"
      >
        <div
          className="navbar-nav"
        >
          <a
            aria-current="page"
            className="nav-link active"
            href="/catalog"
            onClick={[Function]}
            style={Object {}}
          >
            Catalog
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/delivery"
            onClick={[Function]}
          >
            Delivery
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/contacts"
            onClick={[Function]}
          >
            Contacts
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/cart"
            onClick={[Function]}
          >
            Cart
          </a>
        </div>
      </div>
    </div>
  </nav>
  <div
    className="container pt-4"
  >
    <div
      className="Product"
    >
      <div
        className="ProductDetails row"
      >
        <div
          className="col-12 col-sm-5 col-lg-4"
        >
          <img
            className="Image"
            src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkMAYAADkANVKH3ScAAAAASUVORK5CYII="
          />
        </div>
        <div
          className="col-12 col-sm-7 col-lg-6"
        >
          <h1
            className="ProductDetails-Name"
          >
            test-product-name
          </h1>
          <p
            className="ProductDetails-Description"
          >
            test-product-description
          </p>
          <p
            className="ProductDetails-Price fs-3"
          >
            $
            1
          </p>
          <p>
            <button
              className="ProductDetails-AddToCart btn btn-primary btn-lg"
              onClick={[Function]}
            >
              Add to Cart
            </button>
          </p>
          <dl>
            <dt>
              Color
            </dt>
            <dd
              className="ProductDetails-Color text-capitalize"
            >
              black
            </dd>
            <dt>
              Material
            </dt>
            <dd
              className="ProductDetails-Material text-capitalize"
            >
              test-product-material
            </dd>
          </dl>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Catalog:  тест на пустую корзину  1`] = `
<div
  className="Application"
>
  <nav
    className="navbar navbar-expand-sm navbar-light bg-light"
  >
    <div
      className="container"
    >
      <a
        className="Application-Brand navbar-brand"
        href="/"
        onClick={[Function]}
      >
        Example store
      </a>
      <button
        aria-label="Toggle navigation"
        className="Application-Toggler navbar-toggler"
        onClick={[Function]}
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <div
        className="Application-Menu collapse navbar-collapse"
      >
        <div
          className="navbar-nav"
        >
          <a
            aria-current={null}
            className="nav-link"
            href="/catalog"
            onClick={[Function]}
          >
            Catalog
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/delivery"
            onClick={[Function]}
          >
            Delivery
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/contacts"
            onClick={[Function]}
          >
            Contacts
          </a>
          <a
            aria-current="page"
            className="nav-link active"
            href="/cart"
            onClick={[Function]}
            style={Object {}}
          >
            Cart (1)
          </a>
        </div>
      </div>
    </div>
  </nav>
  <div
    className="container pt-4"
  >
    <div
      className="Cart"
    >
      <div
        className="row mb-4"
      >
        <div
          className="col"
        >
          <h1>
            Shopping cart
          </h1>
          <table
            className="Cart-Table table"
          >
            <thead>
              <tr>
                <th
                  scope="col"
                >
                  #
                </th>
                <th
                  scope="col"
                >
                  Product
                </th>
                <th
                  scope="col"
                >
                  Price
                </th>
                <th
                  scope="col"
                >
                  Count
                </th>
                <th
                  scope="col"
                >
                  Total
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                data-testid="0"
              >
                <th
                  className="Cart-Index"
                  scope="row"
                >
                  1
                </th>
                <td
                  className="Cart-Name"
                >
                  test-product-name
                </td>
                <td
                  className="Cart-Price"
                >
                  $
                  1
                </td>
                <td
                  className="Cart-Count"
                >
                  1
                </td>
                <td
                  className="Cart-Total"
                >
                  $
                  1
                </td>
              </tr>
            </tbody>
            <tfoot>
              <tr>
                <td
                  colSpan={4}
                >
                  Order price:
                </td>
                <td
                  className="Cart-OrderPrice"
                >
                  $
                  1
                </td>
              </tr>
            </tfoot>
          </table>
        </div>
      </div>
      <div
        className="row mb-4"
      >
        <div
          className="col-6"
        >
          <button
            className="Cart-Clear btn btn-outline-secondary"
            onClick={[Function]}
          >
            Clear shopping cart
          </button>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-12 col-sm-8 col-md-6 col-lg-4"
        >
          <h2>
            Сheckout
          </h2>
          <div
            className="Form"
          >
            <div
              className="mb-3"
            >
              <label
                className="form-label"
                htmlFor="f-name"
              >
                Name
              </label>
              <input
                autoComplete="off"
                className="Form-Field Form-Field_type_name form-control"
                disabled={false}
                id="f-name"
                onChange={[Function]}
                type="text"
              />
              <div
                className="invalid-feedback"
              >
                Please provide your name
              </div>
            </div>
            <div
              className="mb-3"
            >
              <label
                className="form-label"
                htmlFor="f-phone"
              >
                Phone
              </label>
              <input
                className="Form-Field Form-Field_type_phone form-control"
                disabled={false}
                id="f-phone"
                onChange={[Function]}
                type="text"
              />
              <div
                className="invalid-feedback"
              >
                Please provide a valid phone
              </div>
            </div>
            <div
              className="mb-3"
            >
              <label
                className="form-label"
                htmlFor="f-address"
              >
                Address
              </label>
              <textarea
                className="Form-Field Form-Field_type_address form-control"
                disabled={false}
                id="f-address"
                onChange={[Function]}
                rows={3}
              />
              <div
                className="invalid-feedback"
              >
                Please provide a valid address
              </div>
            </div>
            <button
              className="Form-Submit btn btn-primary"
              disabled={false}
              onClick={[Function]}
            >
              Checkout
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
